app/src/androidTest/java/com/github/ravifrancesco/softwaredevproject/UserPointTest.java

```diff
@@ -0,0 +1,87 @@
+package com.github.ravifrancesco.softwaredevproject;
+
+import android.content.Context;
+import android.content.Intent;
+
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.rule.ServiceTestRule;
+
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.Random;
+import java.util.concurrent.TimeoutException;
+
+
+public class UserPointTest {
+
+    @Rule
+    public final ServiceTestRule serviceRule = new ServiceTestRule();
+
+    @Test
+    public void getLocationTest() throws TimeoutException {
+
+        serviceRule.startService(
+                new Intent(ApplicationProvider.getApplicationContext(),
+                        UserPoint.class));
+
+        Context mContext = ApplicationProvider.getApplicationContext();
+
+        UserPoint userPoint = new UserPoint(mContext);
+
+        if (userPoint.canGetLocation()) {
+            return;
+        } else {
+            Assert.assertEquals(27.988056, userPoint.getLatitude(), 0);
+            Assert.assertEquals(86.925278, userPoint.getLongitude(), 0);
+            Assert.assertEquals(8848.86, userPoint.getAltitude(), 0);
+            Assert.assertEquals(0.0, userPoint.getAccuracy(), 0);
+        }
+
+    }
+
+    @Test
+    public void setLocationTest() throws TimeoutException {
+
+        serviceRule.startService(
+                new Intent(ApplicationProvider.getApplicationContext(),
+                        UserPoint.class));
+
+        Context mContext = ApplicationProvider.getApplicationContext();
+
+        UserPoint userPoint = new UserPoint(mContext);
+
+        Random r = new Random();
+
+        double lat = r.nextDouble();
+        double lon = r.nextDouble();
+        double alt = r.nextDouble();
+        double acc = r.nextDouble();
+
+        if (!userPoint.canGetLocation()) {
+            Assert.assertEquals(27.988056, userPoint.getLatitude(), 0);
+            Assert.assertEquals(86.925278, userPoint.getLongitude(), 0);
+            Assert.assertEquals(8848.86, userPoint.getAltitude(), 0);
+            Assert.assertEquals(0.0, userPoint.getAccuracy(), 0);
+        }
```

Comment:
I think that the test might be a bit weird with this. This test is supposed to test if the function can set it's location. But it will fail if there is something wrong with the default values. The default values are tested in an other test and shouldn't be needed to be tested again.

Reply:
True. My idea was to check if the location is returned correctly at the beginning, then check if the setlocation sets the location correctly, then when the real location is re-enabled it checks if it is still able to get the real location. Does it make sense?

Reply:
I see what you mean. I can just see the issue with that the test could fail because of issues with the default values or that the "switchToRealLocation" function doesn't work. I might be a bit petty, but shouldn't the "switchToRealLocation" have a separate test then instead?

Reply:
The fact is that setLocation and switchToRealLocation are strongly connected